[GStreamer][WebCodecs] Improve support for h.264 frames reordering in decoder

philn opened the pull request:
#### 0745467377affa10f7b1a2aef1a93a148afbe71e
<pre>
[GStreamer][WebCodecs] Improve support for h.264 frames reordering in decoder
<a href="https://bugs.webkit.org/show_bug.cgi?id=265335">https://bugs.webkit.org/show_bug.cgi?id=265335</a>

Reviewed by Youenn Fablet and Xabier Rodriguez-Calvar.

The GStreamer video decoder flush logic was flawed, assuming a flush-start/stop cycle would be
sufficient, but in practice we need to send EOS and teardown the element to force drain of the
decoder.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/webcodecs/videoFrame-canvasImageSource-expected.txt: Copied from LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/videoFrame-canvasImageSource-expected.txt.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/webcodecs/videoFrame-canvasImageSource-expected.txt: Renamed from LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/videoFrame-canvasImageSource-expected.txt.
* Source/WebCore/platform/graphics/gstreamer/VideoDecoderGStreamer.cpp:
(WebCore::GStreamerInternalVideoDecoder::GStreamerInternalVideoDecoder):
(WebCore::GStreamerInternalVideoDecoder::decode):
(WebCore::GStreamerInternalVideoDecoder::flush):
* Source/WebCore/platform/gstreamer/GStreamerElementHarness.cpp:
(WebCore::GStreamerElementHarness::~GStreamerElementHarness):
(WebCore::GStreamerElementHarness::reset):
* Source/WebCore/platform/gstreamer/GStreamerElementHarness.h:

Canonical link: <a href="https://commits.webkit.org/273290@main">https://commits.webkit.org/273290@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2f4d193686560149128b92ec48272bfad781ae05

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34957 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13826 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37015 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37685 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31560 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36107 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16223 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10904 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30498 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/35492 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11720 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31156 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10256 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/10326 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31245 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38937 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31783 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31574 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/36342 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10423 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8341 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34322 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12234 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8016 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10963 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11297 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris reviewed (approved):
Lgtm.

Consider if you want to wait for @youennf to land this or not.

youennf reviewed (approved):
This is ok for the tests as we want to check reordering here, which happens for the first frames.
I would expect though the same decoding behaviour. @philn, do you know how Chromium is behaving on Linux for these tests?

philn committed:
[GStreamer][WebCodecs] Improve support for h.264 frames reordering in decoder
https://bugs.webkit.org/show_bug.cgi?id=265335

Reviewed by Youenn Fablet and Xabier Rodriguez-Calvar.

The GStreamer video decoder flush logic was flawed, assuming a flush-start/stop cycle would be
sufficient, but in practice we need to send EOS and teardown the element to force drain of the
decoder.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/webcodecs/videoFrame-canvasImageSource-expected.txt: Copied from LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/videoFrame-canvasImageSource-expected.txt.
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/webcodecs/videoFrame-canvasImageSource-expected.txt: Renamed from LayoutTests/platform/glib/imported/w3c/web-platform-tests/webcodecs/videoFrame-canvasImageSource-expected.txt.
* Source/WebCore/platform/graphics/gstreamer/VideoDecoderGStreamer.cpp:
(WebCore::GStreamerInternalVideoDecoder::GStreamerInternalVideoDecoder):
(WebCore::GStreamerInternalVideoDecoder::decode):
(WebCore::GStreamerInternalVideoDecoder::flush):
* Source/WebCore/platform/gstreamer/GStreamerElementHarness.cpp:
(WebCore::GStreamerElementHarness::~GStreamerElementHarness):
(WebCore::GStreamerElementHarness::reset):
* Source/WebCore/platform/gstreamer/GStreamerElementHarness.h:

Canonical link: https://commits.webkit.org/273290@main
